### ISSUE TITLE
[C++][FS][Azure] Consider how and if to fix disallowed characters in file metadata

### ISSUE THREAD

**Issue by Tom-Newton** [apache/arrow#40057]
### Describe the enhancement requested

Child of https://github.com/apache/arrow/issues/18014

Manually modifying https://github.com/apache/arrow/pull/40021 to run the Python tests against a real blob storage account caused a failure when attempting to write the metadata `{'Content-Type': 'x-pyarrow/test'}`. Looking at the Generic C++ tests it looks like we will run into the same issue there https://github.com/apache/arrow/blob/c23a097965b5c626cbc91b229c76a6c13d36b4e8/cpp/src/arrow/filesystem/test_util.cc#L912-L913. 

```
E   OSError: CommitBlockList failed for 'https://tomtestflat.blob.core.windows.net/63217166-c9f8-11ee-989a-71cec6336ac8/open-output-stream-metadata'. Committing is required to flush an output/append stream. Azure Error: [InvalidMetadata] 400 The metadata specified is invalid. It has characters that are not permitted.
E   The metadata specified is invalid. It has characters that are not permitted.
E   RequestId:ac61e6da-e01e-0009-2305-5e07de000000
E   Time:2024-02-12T22:45:15.5130135Z
E   Request ID: ac61e6da-e01e-0009-2305-5e07de000000
```

It turns out real Azure storage doesn't allow `-`s in the metadata keys. This behaviour is the same on flat and hierarchical namespace storage accounts but azurite accepts it without error. 

Apparently the keys (names) of metadata must conform to https://learn.microsoft.com/en-us/rest/api/storageservices/naming-and-referencing-containers--blobs--and-metadata#metadata-names the naming rules for [C# identifiers](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference).

We need to decide what we want to do here. I think either we have to accept these limitations or we will need to encode the metadata keys before writing them to azure then decode when reading back. A quick Google search came up with these options for potential encodings https://stackoverflow.com/questions/32037525/encode-to-alphanumeric-in-javascript#:~:text=To%20encode%20to%20an%20alphanumeric,in%20a%20shorter%20encoded%20string.

The downside of encoding the metadata would be that other Azure clients won't know to decode. 

### Component(s)

C++

**Comment by Tom-Newton** [comment]
I think probably this blocks https://github.com/apache/arrow/issues/39069

**Comment by Tom-Newton** [comment]
I like the idea of disalowing `_`. That makes handling `-` much simpler and its more interpretable.

**Comment by pitrou** [comment]
So, the question: which convention do other Azure clients use? What is the recommended or official way to set a file's content-type?

**Comment by Tom-Newton** [comment]
> What is the recommended or official way to set a file's content-type?

This is an interesting point. If we only care about setting specific things like `content-type` then that simplifies the problem. Azure actually separates some values including content type from the user set metadata. I think https://github.com/apache/arrow/pull/40671#pullrequestreview-1948247090 should resolve this.

**Comment by kou** [comment]
Resolved by #40671.
